world_positioning/src/worldP_ekf.py

```diff
@@ -0,0 +1,385 @@
+
+    return euler
+
+def quat_msg_to_euler(q):
+    q_array = quat_msg_to_array(q)
+    return vec_to_msg(quat_to_euler(q_array))
+
+def quat_to_euler_jacobian(q):
+    return jacobian(quat_to_euler(q))
+
+class EKF:
+    def __init__(self):
+        # https://en.wikipedia.org/wiki/Extended_Kalman_filter
+        
+        ''' Model Description
+        x = state
+        u = inputs (sub position and orientation, object position in camera)
+        z = outputs (sub and object position)
+        P = unvertainty/variance matrix of state x
+
+        Dynamics: x[k+1] = f(x[k], u[k])
+        Outputs:    z[k] = h(x[k], u[k])
+
+        Linear Form:
+          x[k+1] = A*x[k] + B*u[k]
+            z[k] = C*x[k] + B*u[k]
+        '''
+
+        ''' Process Description
+
+        Each loop will do a prediction step based on the sub state,
+        to update the expected sub and world object positions and their variance.
+        
+        Then a measurement step, where it uses input from the zed camera's
+        detected objects to refine the prediction and varience
+        
+        Then there is a step where the new expected sub state and world state is converted
+        to a WorldState message which is then published.
+
+        '''
+
+        #first version: only 1 object in state
+
+        self.nState = 13 #number of elements in state
+
+        #n and m are only for 1 point
+        self.n = self.nState+3 #substate + world x,y,z of object
+        self.m = self.nState+3 #sub state + x,y,z of object in camera
+
+        self.x = np.zeros((self.n,1))
+        self.x[IDx.Ow] = 1
+
+        self.u = np.zeros((self.m, 1))
+
+        self.B = np.array() #input modification matrix
+
+        self.P = np.eye(self.n)
+        self.Q = np.eye(self.n) * 0.01 #uncert in dynamics model
+
+        self.camera_info = ZedCamListener() 
+        self.state_info = SubStateListener()
+
+        self.state_msg = SubState()
+        self.camera_msg = Camera() #need to implement this for camera
+        self.world_msg = WorldState()
+
+        self.rate = 20
+        self.rate_ctrl = rospy.Rate(self.rate)
+
+        self.calc_F = jacobian(self.f)
+
+        #self.world_pos_msg = NotImplemented #implement worldpositioning message
+        self.world_pub = rospy.Publisher("world state", NotImplemented, queue_size = 1)
+        #should be in listener already
+        #self.sub_state_sub = rospy.Subscriber("state estimation", NotImplemented, NotImplemented)
+
+        self.last_prediction_t = self.get_t()
+
+    def get_t(self):
+        return rospy.Time.now().to_sec()
+
+    def prediction(self):
+        #jacobian of F
+        F = self.calc_F(self.x, self.u)
+        F = np.reshape(F,(self.n ,self.n))
+        Fx = F[0:self.n, 0:self.n]
+        print("Fx", Fx)
+
+        #update x and P
+        self.x = self.f(self.x, self.u)  #where is this f coming from
+        inter = np.dot(Fx, self.P)
+        self.P = np.dot(inter, np.transpose(Fx)) + self.Q
```

Comment:
You can use np.linalg.multidot([]) here as well for efficiency/readability

Reply:
so it'll look like this? 
self.P = np.linalg.multidot([Fx, self.P, np.transpose(Fx)])+self.Q

Reply:
Yeah exactly. I think it makes it a bit more readable since you dont need inter anymore. Plus numpy can do some optimizations based on the order that it does the dot product in.